deliver/lib/deliver/upload_screenshots.rb

```diff
@@ -52,25 +55,178 @@ def upload(options, screenshots, max_n_threads = 16)
+
+            updated_screenshots_per_device_type[device_type] << {
+                screenshot: screenshot_with_checksum[:screenshot],
+                position: index
+            }
+          end
+        end
+
+        updated_sets_per_language[language] = updated_screenshots_per_device_type
+
+        count = updated_screenshots_per_device_type.values.reduce(0) { |sum, screenshots_with_positions| sum + screenshots_with_positions.size }
+        UI.message("Found #{count} updated screenshots for language #{language}")
+      end
+
+      updated_sets_per_language
+    end
+
+    def get_checksums_to_delete(localizations, updated_sets_per_language)
+      checksums_to_delete = {} # per language and device type
+
+      updated_sets_per_language.each do |language, updated_screenshots_per_device_type|
+        checksums_per_device_type = {}
+        localization = localizations.find do |l|
+          l.locale == language
+        end
+
+        unless localization
+          UI.error("Couldn't find localization on version for #{language}")
+          next
+        end
+
+        app_screenshot_sets = localization.get_app_screenshot_sets
+        app_screenshot_sets.each do |app_screenshot_set|
+          device_type = app_screenshot_set.screenshot_display_type
+
+          unless updated_screenshots_per_device_type.key?(device_type)
+            # if there is no device type specified, add empty array of checksums to delete
+            checksums_per_device_type[device_type] = {
+                checksums: [],
+                count_after_delete: app_screenshot_set.app_screenshots.size
+            }
+            next
+          end
+
+          updated_screenshots_set = updated_screenshots_per_device_type[device_type]
+          updated_screenshots_positions = updated_screenshots_set.map { |screenshot_with_position| screenshot_with_position[:position] }
+          checksums = []
+
+          app_screenshot_set.app_screenshots.each_with_index do |screenshot, index|
+            next unless updated_screenshots_positions.include?(index)
+            checksums << screenshot.source_file_checksum
+          end
+
+          checksums_per_device_type[device_type] = {
+              checksums: checksums,
+              count_after_delete: app_screenshot_set.app_screenshots.size - checksums.size
+          }
+        end
+
+        checksums_to_delete[language] = checksums_per_device_type
+      end
+
+      checksums_to_delete
+    end
+
+    def get_expected_count_after_delete(checksums_to_delete)
+      sum = 0
+
+      checksums_to_delete.values.each do |checksums_per_device_type|
+        checksums_per_device_type.values.each do |checksums_with_count|
+          sum += checksums_with_count[:count_after_delete]
+        end
+      end
+
+      sum
+    end
+
+    def delete_screenshots(localizations, checksums_to_delete, max_n_threads, tries: 5)
```

Comment:
We should make sure that we keep compatibility with the `:overwrite_screenshots` option (which should delete everything)

Reply:
I decided to leave it like this due to the disregard of `overwrite_screenshots`. If we decide this is not the way we want to do stuff, I will update it. 😃